--cfg=hydra for read-only top-level config

jieru-hu opened the pull request:
Closes #1461

omry commented on hydra/_internal/hydra.py:
```diff
@@ -136,13 +136,14 @@ def multirun(
-        with open_dict(cfg.hydra):
-            del cfg.hydra["hydra_help"]
-            del cfg.hydra["help"]
+        with read_write(cfg):
+            with open_dict(cfg):
+                for key in list(cfg.keys()):
+                    if key != "hydra":
+                        del cfg[key]
```
This would a bit more compact:
```suggestion
       with flag_override(["struct", "readonly"], False):
          for key in list(cfg.keys()):
            if key != "hydra":
              del cfg[key]
```

omry commented on tests/test_hydra.py:
```diff
@@ -1227,10 +1227,50 @@ def test_hydra_main_without_config_path(tmpdir: Path) -> None:
+    "overrides,expected",
+    [
+        (
+            ["--cfg", "job"],
```
To make this more compact, you can get a specific value: `-p baud_rate`

omry commented on tests/test_hydra.py:
```diff
@@ -1227,10 +1227,50 @@ def test_hydra_main_without_config_path(tmpdir: Path) -> None:
+            ),
+        ),
+        (
+            ["--cfg", "hydra", "-p", "hydra.env"],
```
same here. for example `-p hydra.job.name`.

omry reviewed (commented):
I commented on the original issue with a few related problems, can you check that those are covered as well?

jieru-hu commented:
>I commented on the original issue with a few related problems, can you check that those are covered as well?

sorry missed the comment. added fix and test for help and info.


> This would a bit more compact:

nice!


> To make this more compact, you can get a specific value: -p baud_rate

fixed!

omry commented on tests/test_hydra.py:
```diff
@@ -1227,10 +1227,31 @@ def test_hydra_main_without_config_path(tmpdir: Path) -> None:
+@mark.parametrize(
+    "overrides,expected",
+    [
+        (["--cfg", "job", "-p", "baud_rate"], "19200"),
+        (["--cfg", "hydra", "-p", "hydra.job.name"], "frozen"),
+        (["--info", "config"], "baud_rate: 19200"),
+        (["--hydra-help"], "== Flags =="),
+        (["--help"], "frozen is powered by Hydra."),
```
neat :)
